### ISSUE TITLE
브랜치 전략: feature-model-backbones

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#11]
## 목표와 범위
- **목표:**
    프로젝트에 사용할 **백본 모델군의 선택·튜닝·프로파일링**을 표준화하여, **성능–속도–메모리** 관점의 균형점을 정량적으로 제시합니다.
    
    - 본 브랜치는 **timm 기반 백본 팩토리/레지스트리, 권장 입력 해상도 매핑, 헤드 교체, PosEmbed 보간(ViT), DropPath/Dropout, Freezing/Unfreezing 스케줄, 지표·속도·메모리·FLOPs 벤치마크, 모델 카드/리포트**를 제공합니다.
    - 모든 기능은 **`ipynb`에서 사전 검증** 후, 통과 코드만 `src/`의 표준 디렉터리로 **모듈화 이관**합니다.
    - 병합 전 **성능표·지연시간·메모리 사용량·DoD 기준**을 충족하는 것을 완료 기준(DoD)으로 정의합니다.
- **범위:**
    - **지원 백본(초기 세트):** `resnet34/50/101`, `efficientnet_b0/b3/b4`, `convnext_tiny/base`, `vit_base_patch16_224`, `swin_tiny_patch4_window7_224`
    - **입력 해상도 매핑:** ResNet/ConvNeXt 224(옵션 288), EffNet-B3 300(옵션 288/320), B4 380, ViT 224(옵션 384), Swin-T 224(옵션 256)
    - **표준화 항목:** 분류 헤드 교체·가중치 초기화, 글로벌 풀/레이어노름, Dropout/DropPath, PosEmbed 보간, AMP/bfloat16 호환, GradCheckpoint(옵션)
    - **프로파일링:** 파라미터 수·FLOPs(해상도별)·Latency(imgs/sec)·메모리 피크, GPU/CPU 공통 인터페이스
    - **리포팅:** 모델 카드(요약), 모델 ZOO 표(CSV/HTML)

---

### 1) 백본 팩토리·레지스트리
- **팩토리:** `build_model(arch:str, num_classes:int, img_size:int, drop:float=0.0, drop_path:float=0.0, pretrained:bool=True, global_pool:str="avg")`
    - 내부적으로 `timm.create_model` 사용, `num_classes` 전달 또는 헤드 수동 교체
    - **ViT/Swin 해상도 변경 시**: Positional Embedding 보간(비율 기반), 윈도우 크기 점검(불일치 시 에러)
    - **Classifier Head**: `nn.Linear(in_features, 17)` + 선택 Dropout, Kaiming/Xavier 초기화 옵션
    - **BN/LayerNorm Freezing 옵션**: `freeze_stages`, `bn_eval`, `bn_frozen`
    - **AMP/bfloat16** 대응: forward에서 dtype 가이드(훈련 루프에서 관리)
- **레지스트리:** `MODEL_REGISTRY`
    - 키: `arch` 문자열, 값: 메타정보(dict) – 권장 img_size, family, params_ref, notes
    - **구성 파일**: `configs/backbones_registry.json` (예시는 “구성·설정 표준” 참조)

### 2) 권장 해상도·하이퍼 매핑
- **기본 매핑(초기 권장):**
    - ResNet50: 224, DropPath 0.0~0.1
    - EffNet-B3: 300, Dropout 0.2, DropPath 0.1
    - EffNet-B4: 380, Dropout 0.3, DropPath 0.2
    - ConvNeXt-T: 224(옵션 288), DropPath 0.1~0.2
    - ViT-B/16: 224(옵션 384), Dropout 0.1, DropPath 0.1
    - Swin-T: 224(옵션 256), DropPath 0.1
- **Freezing/Unfreezing 스케줄(권장):**
    - Warmup 1~2 epoch 동안 **Stem+Stage1** Freeze → 이후 전층 Finetune
    - 또는 **헤드만 1 epoch** → 전체 해제(학습 안정성 우선)

### 3) 프로파일링·호환 API
- **프로파일러 모듈:** `profile_model(arch, img_size, batch_size, device)`
    - 산출: `params(M)`, `flops(G)`, `latency(imgs/sec)`, **peak_mem(MB)**
    - FLOPs: 입력 해상도 기반 추정(내부 연산 카운터), Conv/Linear/Attention 지원
    - Latency: 워밍업 N=50, 측정 M=200 iter 평균, AMP on/off 분리 기록
- **Shape/정합 검사:** 랜덤 텐서 입력으로 forward 체크, 출력 shape=[B,17] 보장
- **에러 가드:** 입력 크기와 모델 요구사항 충돌 시 명시적 예외 메시지

### 4) 모델 카드·모델 ZOO
- **모델 카드(자동 생성):** 각 백본별
    - 요약(해상도/파라미터/DropPath/학습시간/최종 val Macro-F1)
    - 장단점(속도 vs 성능), 권장 사용처(단일/앙상블 역할)
- **모델 ZOO 표:** `experiments/backbones/model_zoo.csv`
    - 컬럼: `arch,img_size,params_M,flops_G,imgs_per_sec,peak_mem_MB,macro_f1_clean,macro_f1_cv,notes`

### 5) 학습 루틴 연계(훈련/검증)
- **`feature-baseline-training` 연동:**
    - 백본 이름만 바꿔 **동일 파이프라인**으로 학습 가능(증강/옵티마 동일)
    - 클래스 가중치·WeightedSampler 옵션 유지
- **`feature-eval-cv` 연동:**
    - 폴드별 예측 수집·집계 자동화(메트릭/리포트)
- **`feature-robust-optim` 연동:**
    - DropPath/EMA/SAM 등 강건 옵션 스위치 반영

### 6) 추론·내보내기(옵션)
- **추론 엔진 헬퍼:** `src/inference/infer_single_model.py`에서 백본 로드/전처리/추론
- **ONNX/TorchScript 내보내기**(선택): `src/models/export.py` – 입력 크기 고정 시만 허용
- **정밀도 전환**: FP16/BF16 추론 플래그

---

## ipynb-우선 워크플로우 (파일명·셀 단위 체크리스트)
1. **`notebooks/BACKBONES/00_backbones_registry_and_configs.ipynb`**
    - [ ]  `configs/backbones_registry.json` 작성(권장 해상도/드랍/노트)
    - [ ]  팩토리/레지스트리 import·Smoke Test(각 백본 1회 forward)
2. **`notebooks/BACKBONES/01_profile_params_flops_latency.ipynb`**
    - [ ]  각 백본 × 해상도 조합에 대한 `params/flops/imgs_sec/peak_mem` 측정
    - [ ]  `model_zoo.csv` 생성
3. **`notebooks/BACKBONES/02_single_fold_training_sanity.ipynb`**
    - [ ]  백본 3종(ResNet50/EffNet-B3/ConvNeXt-T) 단일 폴드 10~15epoch 러프 학습
    - [ ]  val Macro-F1, 수렴 곡선 비교
4. **`notebooks/BACKBONES/03_kfold_training_and_cv_report.ipynb`**
    - [ ]  선택 백본 K=5 학습 → `feature-eval-cv`로 리포트 생성
    - [ ]  평균/표준편차/CI 기록
5. **`notebooks/BACKBONES/04_vit_swin_resolution_ablation.ipynb`**
    - [ ]  ViT/Swin 해상도(224↔384/256) 비교, PosEmbed 보간·윈도우 규칙 검증
    - [ ]  DropPath 스윕(0.0/0.1/0.2)
6. **`notebooks/BACKBONES/05_ensemble_candidate_selection.ipynb`**
    - [ ]  상보성 분석(혼동 상위 클래스 개선 여부)로 앙상블 후보 선정
    - [ ]  추후 `feature-infer-ens-submit` 연계 메모 저장
7. **`notebooks/BACKBONES/06_export_model_modules.ipynb`**
    - [ ]  통과 코드만 `src/models/*.py`, `src/training/*.py`로 이관
    - [ ]  `__all__`/docstring/타입힌트/사용 예시 정리

---

## 산출물(Artifacts)
- **코드 모듈**
    - `src/models/backbone.py` : 팩토리/레지스트리, 헤드 교체, PosEmbed 보간, freeze 옵션
    - `src/models/profiling.py` : 파라미터·FLOPs·Latency·메모리 측정 유틸
    - `src/models/utils.py` : DropPath/Init/Norm 유틸, GradCheckpoint 옵션
    - (옵션) `src/models/export.py` : TorchScript/ONNX 내보내기
- **리포트/로그**
    - `experiments/backbones/model_zoo.csv`
    - `experiments/backbones/<run_id>/{profile.json, train_log.csv, curves.png, summary.json}`
    - `docs/backbones_report_<timestamp>.html` (모델 카드 묶음 리포트)
- **설정 파일**
    - `configs/backbones_registry.json`
- **노트북(최종 파일명 확정)**
    - `notebooks/BACKBONES/00_backbones_registry_and_configs.ipynb`
    - `notebooks/BACKBONES/01_profile_params_flops_latency.ipynb`
    - `notebooks/BACKBONES/02_single_fold_training_sanity.ipynb`
    - `notebooks/BACKBONES/03_kfold_training_and_cv_report.ipynb`
    - `notebooks/BACKBONES/04_vit_swin_resolution_ablation.ipynb`
    - `notebooks/BACKBONES/05_ensemble_candidate_selection.ipynb`
    - `notebooks/BACKBONES/06_export_model_modules.ipynb`

---

### 1 백본 빌드·호환성
- `build_model()`은 모든 지원 아키텍처에서 **출력 shape=[B,17]** 보장.
- `img_size` 변화 시 ViT PosEmbed 보간·Swin 윈도우 호환 체크 → 불일치 시 명시적 예외.
- `global_pool`/`drop`/`drop_path`/`freeze_stages`/`bn_frozen` 인자 지원.

### 2 프로파일링
- 입력: `arch,img_size,batch_size,device,amp_flag`
- 출력: `params_M,flops_G,imgs_sec,peak_mem_MB`
- 워밍업/측정 스텝 수 고정, 결과 JSON/CSV 저장.

### 3 학습 연계
- `feature-baseline-training`의 훈련 루프와 **완전 호환**(옵티마/스케줄 동일).
- `feature-eval-cv`와 연동해 폴드별 예측/지표 수집.

### 4 리포팅
- 백본별 모델 카드 자동 생성, ZOO 표와 함께 HTML 리포트 생성.
- 권장 사용 시나리오(단일/앙상블·고속/고성능) 명시.

---

## 실험·검증(Validation) 기준
| 항목 | 측정/검증 방법 | 통과 기준(제안) |
| --- | --- | --- |
| **호환성** | 모든 백본 smoke forward | 출력 shape 정합, 에러 0건 |
| **프로파일링 신뢰성** | 반복 측정(3회) | imgs/sec 변동 ≤ **±5%** |
| **기본 성능** | 단일 폴드 val Macro-F1 | ResNet50 224 기준 **베이스라인 이상** |
| **CV 일관성** | K=5 평균±표준편차 | 표준편차 ≤ **±2.0%p** |
| **리포트** | HTML/CSV 산출 | 필수 섹션/표/그림 포함, 링크 정상 |
| **재현성** | 동일 설정 2회 | Macro-F1 편차 ≤ **±1.0%p** |
---

## 병합 조건(Definition of Done & PR 체크리스트)
- **DoD**
    - ✅ 00~06 노트북 **모든 셀 실행 완료(오류 0, 출력 포함)**
    - ✅ `model_zoo.csv` 생성 및 최소 5개 백본의 프로파일링/단일 폴드 성능 기록
    - ✅ HTML 리포트 1부(`docs/backbones_report_*.html`) 생성
    - ✅ `src/models/backbone.py` 팩토리/레지스트리 완성, shape/PosEmbed 가드 포함
    - ✅ `feature-baseline-training`/`feature-eval-cv` 연동 검증
- **PR 체크리스트**
    - [ ]  코드 포맷팅(black/isort), 린트 통과
    - [ ]  단위 테스트 4종 통과(팩토리/forward/PosEmbed/프로파일러)
    - [ ]  `configs/backbones_registry.json` 커밋
    - [ ]  아티팩트 링크(모델 ZOO·리포트·로그) 첨부
    - [ ]  규칙 준수 확인(외부 데이터·PL·유료 API 미사용)

---

## 브랜치 작업 순서(권장 커밋 플로우)
1. `feat(registry)`: backbones_registry.json·팩토리 스켈레톤
2. `feat(profile)`: profiling 유틸 및 01 노트북
3. `feat(resnet/effnet/convnext)`: 3종 우선 지원·훈련 스모크
4. `feat(vit/swin)`: PosEmbed/윈도우 가드·해상도 어블레이션
5. `feat(zoo)`: model_zoo.csv·리포트 생성
6. `feat(export)`: 모듈 이관(06)
7. `tests/*`: 단위 테스트 확충

---

## 구성·설정 표준
**`configs/backbones_registry.json` (예시)**

```json
{
  "resnet50":   { "family": "resnet",   "default_img_size": 224, "drop": 0.0, "drop_path": 0.0, "notes": "fast, solid baseline" },
  "efficientnet_b3": { "family": "efficientnet", "default_img_size": 300, "drop": 0.2, "drop_path": 0.1, "notes": "good trade-off" },
  "efficientnet_b4": { "family": "efficientnet", "default_img_size": 380, "drop": 0.3, "drop_path": 0.2, "notes": "higher accuracy" },
  "convnext_tiny": { "family": "convnext", "default_img_size": 224, "drop": 0.0, "drop_path": 0.1, "notes": "modern cnn" },
  "vit_base_patch16_224": { "family": "vit", "default_img_size": 224, "drop": 0.1, "drop_path": 0.1, "notes": "needs data/regularization" },
  "swin_tiny_patch4_window7_224": { "family": "swin", "default_img_size": 224, "drop": 0.0, "drop_path": 0.1, "notes": "window attention" }
}

```

---

## 파라미터·해상도 표준(요약)
| 백본 | 권장 해상도 | Dropout | DropPath | 비고 |
| --- | --- | --- | --- | --- |
| ResNet50 | 224 | 0.0 | 0.0~0.1 | 빠른 기준선 |
| EffNet-B3 | 300 | 0.2 | 0.1 | 균형형 |
| EffNet-B4 | 380 | 0.3 | 0.2 | 고정확도(리소스↑) |
| ConvNeXt-T | 224(→288) | 0.0 | 0.1~0.2 | 현대 CNN |
| ViT-B/16 | 224(→384) | 0.1 | 0.1 | PosEmbed 보간 필요 |
| Swin-T | 224(→256) | 0.0 | 0.1 | 윈도우 규칙 확인 |
---

## 실패 정책(Failure Policy)
- **호환성 실패**(해상도/윈도우/PosEmbed): 즉시 실패, 가이드 메시지 출력(권장 해상도 표시).
- **수렴 실패/NaN**: 해당 백본 제외 또는 DropPath/Dropout/해상도 완화 후 재시도.
- **프로파일링 누락**: `model_zoo.csv` 필수 컬럼 누락 시 머지 불가.
- **리포트 불완전**: 필수 표/그림 미포함 시 머지 불가.

---

## 리스크·준수사항
- **규칙 준수:** 외부 데이터/테스트 데이터 활용 금지(PL 금지), 유료 API 미사용.
- **리소스 리스크:** EffNet-B4/고해상도는 메모리·시간 급증 → AMP/Accumulation, 배치 조절.
- **일관성 리스크:** 백본별 권장 해상도 상이 → `backbones_registry.json` 단일 소스로 통제.
- **과적합 리스크:** ViT/Swin은 데이터 요구량↑ → DropPath/강건 최적화 레시피와 병행.
